### PR TITLE
Add mean field node to basic tests

### PR DESCRIPTION
**Pull request by TheoCabannes**
This should enable using basic tests in mean field games.

Currently the tests are failing for:
- crowd_modelling_2d_test (Child aborted) (I do not know why).
- crowd_modelling_test.py and predator_prey_test.py due to `pyspiel.SpielError: MaxChanceNodesInHistory() is not implemented`. I do not understand why this issue is not raised in other python tests that call `random_sim_test`.
I am submitting this PR to get some help on it.

### CODE REVIEW

**Comment by RaphaelMarinier** [comment]
Regarding crowd_modelling_2d_test, I have a high suspicion that this is caused by a bug when computing the observation tensor. Those bugs were fixed in the 1D game [here](https://github.com/deepmind/open_spiel/commit/052d88ca99063834cf8fe70bec3a1bdaa2d5ed45), but are still present in the 2D version. I just fixed them internally.

Regarding MaxChanceNodesInHistory, I'll have a look why you get this error with `MaxChanceNodesInHistory()`.

**Comment by lanctot** [comment]
Hello @TheoCabannes, if you check the logs it seems that the errors have the same root cause:

```
MaxChanceNodesInHistory() not implemented.
```

It is referring to this function: https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/spiel.h#L859

Do MFGs have chance nodes? If so, we should implement the function for the MFGs. If not, can we we do a similar `return 0` here?

**Comment by TheoCabannes** [comment]
Hello @lanctot, thank you a lot.
I have identify this function. I have also identify the binding in python https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/python/pybind11/pyspiel.cc#L351

However, what I do not understand is that `MaxChanceNodesInHistory() not implemented.` does not seem to be an issue for [python games](https://github.com/deepmind/open_spiel/tree/master/open_spiel/python/games): https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/python/games/iterated_prisoners_dilemma_test.py#L59 (btw just spotted a nit in the name of the test method `test_game_from_cc` should probably be `test_random_sim_game`) while `MaxChanceNodesInHistory() not implemented.` is raised in the mfg python games (c.f. the tests). 
I do not know if there is some binding issue in mfg python games for `MaxChanceNodesInHistory()` or if `MaxChanceNodesInHistory()` is never called for python game in random sim but is called for mfg python games in [RamdonSimulation](https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/tests/basic_tests.cc#L258).

**To make a long story short, I was not able to understand why I got this issue in the tests (and why it is not an issue for 'regular' python games). I think this might be an issue in the binding.**

**Comment by TheoCabannes** [comment]
> Do MFGs have chance nodes? If so, we should implement the function for the MFGs. If not, can we we do a similar `return 0` here?
MFGs have chance nodes, should `max_chance_nodes_in_history` functions be implemented in python mfgs? Also it seems that `MaxChanceNodesInHistory` is not implemented for C++ mfgs (therefore I suspect that there is a default implementation for MFGs in C++ that might not exists in python).

**Comment by lanctot** [comment]
I spoke to @perolat, he says we should implement that function in the MFG games. 

There may be other tests that fail after that, and we can handle them one-by-one as the tests fail.

Can you take charge of this @TheoCabannes  ?

**Comment by lanctot** [comment]
There is a basic/empty/stub implementation in the base class, but it needs to be overridden by each game individually like the other methods. 

The error is getting triggered within the base class's default implementation, here: https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/spiel.h#L863

**Comment by lanctot** [comment]
> MFGs have chance nodes. Should `max_chance_nodes_in_history` functions be implemented in python mfgs?

Yes, if we want to use the standard tests, they have to be. Same with the C++ MFGs, they each have to override `Game::MaxChanceNodesInHistory`.

And there may be other functions in this category once these are fixed. But probably best to catch them now before there are too many MFGs.

**Comment by TheoCabannes** [comment]
Sure I can look at that. I will ask on mfg support chat room for more direction.

However I do not understand why the tests are passing for '#49 - crowd_modelling_test', MaxChanceNodesInHistory is not implemented in C++ crowd modelling but the test is passing.

**Comment by lanctot** [comment]
> However I do not understand why the tests are passing for '#49 - crowd_modelling_test', MaxChanceNodesInHistory is not implemented in C++ crowd modelling but the test is passing.

Ah, yes it is implemented in that game (in the header file because it's small): https://github.com/deepmind/open_spiel/blob/9e4f6d219244dcd3deddf1ca75ab1113580bc73c/open_spiel/games/mfg/crowd_modelling.h#L134

**Comment by TheoCabannes** [comment]
Thank you. My bad, I did not see it. Sorry about that.

**Comment by lanctot** [comment]
> Thank you. My bad, I did not see it. Sorry about that.

No worries! It's good, because it provides an example. Also, because it is passing, maybe that means this is the only function that needs to be done before all the MFGs can move to the standard tests.

**Comment by TheoCabannes** [comment]
I have added max_chance_nodes_in_history but the tests are failing on my computer due to https://github.com/deepmind/open_spiel/issues/657

**Comment by TheoCabannes** [comment]
Converting to drafts as more bugs seems to exist. I will do another PR for crowd modelling C++ and another one for MaxChanceNodesInHistory binding.

**Comment by lanctot** [comment]
Ok, great. But feel free to reuse this PR for everything related to these.

**Comment by TheoCabannes** [comment]
@RaphaelMarinier what should be `max_chance_nodes_in_history` for the `MFGPredatorPreyGame`?
I put `horizon` but the tests are failing (you should be able to see the tests results soon).

Also, for MFGPredatorPreyGame test with random sim I changed the signature of `RandomSimTest` to :
```
void RandomSimTest(
    const Game& game, int num_sims,
    bool serialize = true, bool verbose = true, bool mask_test = true,
    const std::function<void(const State&)>& state_checker_fn
        = &DefaultStateChecker,
    int mean_field_population = -1);
```

**Comment by lanctot** [comment]
I believe @RaphaelMarinier is on holiday. @perolat is telling me it should be like the other games, i.e. `horizon + 1`.

File: open_spiel/python/mfg/games/predator_prey.py [hunk 1]
```diff
@@ -116,6 +116,10 @@ def new_initial_state(self):
     """
     return MFGPredatorPreyState(self)
 
+  def max_chance_nodes_in_history(self):
+    """Maximun chance nodes in game history."""
+    return self.horizon
```
Comment (lanctot) [inline open_spiel/python/mfg/games/predator_prey.py:121]: Try horizon + 1?

**Comment by TheoCabannes** [comment]
Closing this PR as https://github.com/deepmind/open_spiel/pull/661 should solve the issue of predator prey game and https://github.com/deepmind/open_spiel/pull/662 has been created to solve the issue for crowd modelling 2d game.

**Pull request closed without merging**